packages/block-editor/src/components/block-list/use-multi-selection.js

```diff
@@ -0,0 +1,200 @@
+			multiSelectedBlockClientIds: getMultiSelectedBlockClientIds(),
+			hasMultiSelection: hasMultiSelection(),
+			getBlockParents,
+		};
+	}
+
+	const {
+		blockClientIds,
+		isSelectionEnabled,
+		isMultiSelecting,
+		multiSelectedBlockClientIds,
+		hasMultiSelection,
+		getBlockParents,
+	} = useSelect( selector );
+	const {
+		startMultiSelect,
+		stopMultiSelect,
+		multiSelect,
+		selectBlock,
+	} = useDispatch( 'core/block-editor' );
+	const rafId = useRef();
+	const startClientId = useRef();
+
+	/**
+	 * When the component updates, and there is multi selection, we need to
+	 * select the entire block contents.
+	 */
+	useEffect( () => {
+		if ( ! hasMultiSelection || isMultiSelecting ) {
+			return;
+		}
+
+		const { length } = multiSelectedBlockClientIds;
+		// These must be in the right DOM order.
+		const start = multiSelectedBlockClientIds[ 0 ];
+		const end = multiSelectedBlockClientIds[ length - 1 ];
+		const startIndex = blockClientIds.indexOf( start );
+
+		// The selected block is not in this block list.
+		if ( startIndex === -1 ) {
+			return;
+		}
+
+		let startNode = ref.current.querySelector(
+			`[data-block="${ start }"]`
+		);
+		let endNode = ref.current.querySelector(
+			`[data-block="${ end }"]`
+		);
+
+		const selection = window.getSelection();
+		const range = document.createRange();
+
+		// The most stable way to select the whole block contents is to start
+		// and end at the deepest points.
+		startNode = getDeepestNode( startNode, 'start' );
+		endNode = getDeepestNode( endNode, 'end' );
+
+		range.setStartBefore( startNode );
+		range.setEndAfter( endNode );
+
+		selection.removeAllRanges();
+		selection.addRange( range );
+	}, [
+		hasMultiSelection,
+		isMultiSelecting,
+		multiSelectedBlockClientIds,
+		blockClientIds,
+		selectBlock,
+	] );
+
+	const onSelectionChange = useCallback( () => {
+		const selection = window.getSelection();
+
+		// If no selection is found, end multi selection.
+		if ( ! selection.rangeCount || selection.isCollapsed ) {
+			return;
+		}
+
+		let { focusNode } = selection;
+		let clientId;
+
+		// Find the client ID of the block where the selection ends.
+		do {
+			focusNode = focusNode.parentElement;
+		} while (
+			focusNode &&
+			! ( clientId = focusNode.getAttribute( 'data-block' ) )
+		);
+
+		if ( startClientId.current === clientId ) {
+			selectBlock( clientId );
+		} else {
+			const startPath = [ ...getBlockParents( startClientId.current ), startClientId.current ];
+			const endPath = [ ...getBlockParents( clientId ), clientId ];
+			const depth = Math.min( startPath.length, endPath.length ) - 1;
+
+			multiSelect( startPath[ depth ], endPath[ depth ] );
+		}
+	}, [ selectBlock, getBlockParents, multiSelect ] );
+
+	/**
+	 * Handles a mouseup event to end the current mouse multi-selection.
+	 */
+	const onSelectionEnd = useCallback( () => {
+		document.removeEventListener( 'selectionchange', onSelectionChange );
+		// Equivalent to attaching the listener once.
```

Comment:
This comment is confusing to me. I'm not sure what exactly it is trying to say.

Reply:
Perhaps it's not really necessary to comment. I see the `mouseup` listener as equivalent to

```js
windew.addEventListener( 'mouseup', listener, { once: true } );
```

because here we remove the listener on the first call.